Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -509,6 +509,97 @@ $ adb install UnCrackable-Level1.objection.apk
 $ objection explore
 ```
 
+#### Basic Information Gathering
+
+As mentioned previously, Android runs on top of a modified Linux kernel and retains the [proc filesystem](https://www.kernel.org/doc/Documentation/filesystems/proc.txt "procfs") (procfs) from Linux, which is mounted at `/proc`. Procfs provides a directory-based view of a process running on the system, providing detailed information about the process itself, its threads, and other system-wide diagnostics. Procfs is arguably one of the most important filesystems on Android, where many OS native tools depend on it as their source of information.
+
+Many command line tools are not shipped with the Android firmware to reduce the size, but can be easily installed on a rooted device using BusyBox. We can also create our own custom scripts using commands like `cut`, `grep`, `sort` etc, to parse the proc filesystem information.
+
+In this section, we will be using information from procfs directly or indirectly to gather information about a running process.
+
+##### Open Files
+
+You can use `lsof` with the flag `-p <pid>` to return the list of open files for the specified process. See the [man page](http://man7.org/linux/man-pages/man8/lsof.8.html "Man Page of lsof") for more options.
+
+```
+# lsof -p 6233
+COMMAND     PID       USER   FD      TYPE             DEVICE  SIZE/OFF       NODE NAME
+.foobar.c  6233     u0_a97  cwd       DIR                0,1         0          1 /
+.foobar.c  6233     u0_a97  rtd       DIR                0,1         0          1 /
+.foobar.c  6233     u0_a97  txt       REG             259,11     23968        399 /system/bin/app_process64
+.foobar.c  6233     u0_a97  mem   unknown                                         /dev/ashmem/dalvik-main space (region space) (deleted)
+.foobar.c  6233     u0_a97  mem       REG              253,0   2797568    1146914 /data/dalvik-cache/arm64/system@framework@boot.art
+.foobar.c  6233     u0_a97  mem       REG              253,0   1081344    1146915 /data/dalvik-cache/arm64/system@framework@boot-core-libart.art
+...
+```
+
+In the above output, the most relevant fields for us are:
+- `NAME`: path of the file
+- `TYPE`: type of the file, example, file is a directory or a regular file
+
+This can be extremely useful in getting list of uncommon files used while dealing with applications using obfuscation, or other anti-reverse engineering techniques. For instance, an application performing encryption-decryption of data and storing it in a file temporarily.
```

Comment:
What do you think about this?
```suggestion
This can be extremely useful to spot unusual files when monitoring applications using obfuscation or other anti-reverse engineering techniques, without having to reverse the code. For instance, an application might be performing encryption-decryption of data and storing it in a file temporarily.
```

Reply:
This is nice!! :)